packages/core/contracts/optimistic-assertor/implementation/examples/Insurance.sol

```diff
@@ -0,0 +1,97 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+
+import "../../interfaces/OptimisticAssertorInterface.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract Insurance {
+    using SafeERC20 for IERC20;
+    IERC20 public immutable defaultCurrency;
+    OptimisticAssertorInterface public immutable oa;
+    uint256 public constant assertionLiveness = 7200;
+
+    struct Policy {
+        uint256 insuranceAmount;
+        address payoutAddress;
+        address insurer;
+        bytes insuredEvent;
+        bool settled;
+    }
+
+    mapping(bytes32 => bytes32) public oaIdentifiers;
+
+    mapping(bytes32 => Policy) public policies;
+
+    event InsuranceIssued(
+        bytes32 indexed policyId,
+        bytes insuredEvent,
+        uint256 insuranceAmount,
+        address indexed payoutAddress,
+        address indexed insurer
+    );
+
+    event InsurancePayoutRequested(bytes32 indexed policyId, bytes32 indexed assertionId);
+
+    event InsurancePayoutSettled(bytes32 indexed policyId, bytes32 indexed assertionId);
+
+    constructor(address _defaultCurrency, address _optimisticAssertor) {
+        defaultCurrency = IERC20(_defaultCurrency);
+        oa = OptimisticAssertorInterface(_optimisticAssertor);
+    }
+
+    function issueInsurance(
+        uint256 insuranceAmount,
+        address payoutAddress,
+        bytes memory insuredEvent
+    ) public returns (bytes32 policyId) {
+        policyId = keccak256(abi.encode(insuredEvent, payoutAddress));
+        require(policies[policyId].insurer == address(0), "Policy already exists");
+        policies[policyId] = Policy({
+            insuranceAmount: insuranceAmount,
+            payoutAddress: payoutAddress,
+            insurer: msg.sender,
+            insuredEvent: insuredEvent,
+            settled: false
+        });
+        defaultCurrency.safeTransferFrom(msg.sender, address(this), insuranceAmount);
+        emit InsuranceIssued(policyId, insuredEvent, insuranceAmount, payoutAddress, msg.sender);
+    }
+
+    function requestPayout(bytes32 policyId) public returns (bytes32 assertionId) {
+        uint256 bond = oa.getMinimumBond(address(defaultCurrency));
+        defaultCurrency.safeTransferFrom(msg.sender, address(this), bond);
+        defaultCurrency.safeApprove(address(oa), bond);
+        assertionId = oa.assertTruthFor(
+            policies[policyId].insuredEvent,
```

Comment:
it might be illustrative to show this with a more complex `claim` object, such as concatenating something like "Insurance contract is claiming that insurance event ${x} is value at time ${blockTime}".

Reply:
If there were multiple policies with the same insured event but different payout addresses only one could be asserted if called by the same requester due to generating the same `assertionId`. Thus it is better to add request timestamp to the claim.

Reply:
I would suggest stamping block timestamp at `requestPayout`, not when issuing insurance